filebeat/input/v2/plugin.go

```diff
@@ -0,0 +1,112 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"github.com/elastic/beats/heartbeat/scheduler"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/go-concert/chorus"
+)
+
+type Plugin struct {
+	Name        string
+	Constraints Constraint
+	Doc         string
+
+	// TODO: config schema info for config validation
+
+	// Configure configures an Input if possible. Returns an error if the
+	// configuration is false.
+	// The input return must not attempt to run or create any kind of connection yet.
+	// The input is only supposed to be a types place holder for the untyped
+	// configuration.
+	Configure func(log *logp.Logger, config *common.Config) (Input, error)
+}
+
+// Input defines the functionality all input plugins must provide.
+type Input interface {
+	InputTester
+	RunnerFactory
+}
```

Comment:
I know I have raised the issue directly but input should probably have some kind of ids, we can make then automatically generated and definable by users.  The later is probably even better to enforce a unique identifier on each input definition, lets say we want to correlate logs events and specific configuration or monitoring. 

For monitoring or reporting events to fleet we can get around that by saving that information in the observer, but that could be useful for log statements too.

Reply:
I wonder where we should enforce the ID. And how to pass it. Maybe it should not be enforced by the plugin API, but the architecture using the plugins would add/manage ids. As the logger and overall context is passed to the plugin, we can have the underlying framework populate this information. In this case the ID could be passed as parameter in the context, in case an input wants to report it.

If the ID is pushed top-down, it is the framework managing IDs. Maybe this could be used together with the scheduler. E.g. if the ID is given (static), then we enable support for storing the schedule/task state in the registry. In this case no the input/plugin developer doesn't need to handle any kind of management info.

Reply:
>  but the architecture using the plugins would add/manage ids. As the logger and overall context is passed to the plugin.

Correct, we would need some kind of bookkeeping of ids especially if we let people define them, to make sure they are indeed unique. 

What if, the ID is a system option (in the yaml) when defining the plugin so plugins author doesn't see it or have to define it. We could make it available to the passed context to the plugin so an author can reference that if needed (identify a client to a remote system maybe?)

> Maybe this could be used together with the scheduler. E.g. if the ID is given (static)

The goal for having the scheduler that information is to recover execution information from last tick()? I am assuming that if an inputs configuration is changed we might want to try to keep the same ID as much as possible to keep correlation possible and allow us to answer the following question.

Did the config changes actually improve or not the performance of the input.